### PR TITLE
feat: Wait for all handled requests to resolve via `.flush()`

### LINKED ISSUE #71: How to wait for all the responses to finish?
Is there anything like `fetchMock.fush()` ?
http://www.wheresrhys.co.uk/fetch-mock/api#flush
This is how they do it:
```
FetchMock.flush = function() {
	return Promise.all(this._holdingPromises);
};
```
```
FetchMock.fetchHandler = function(url, opts) {
	const response = this.executeRouter(url, opts);

	[...]

	// this is used to power the .flush() method
	let done;
	this._holdingPromises.push(new this.config.Promise(res => (done = res)));

	// wrapped in this promise to make sure we respect custom Promise
	// constructors defined by the user
	return new this.config.Promise((res, rej) => {
		this.generateResponse(response, url, opts)
			.then(res, rej)
			.then(done, done);
	});
};
```

### PR DESCRIPTION
**Pull request by offirgolan**
Resolves #71.

### CODE REVIEW

File: packages/@pollyjs/core/src/polly.js [hunk 1]
```diff
@@ -234,6 +234,14 @@ export default class Polly {
     }
   }
 
+  async flush() {
+    const NOOP = () => {};
```
Comment (jasonmit) [inline packages/@pollyjs/core/src/polly.js]: Whats the roll of the noops here

File: packages/@pollyjs/core/src/polly.js [hunk 2]
```diff
@@ -234,6 +234,14 @@ export default class Polly {
...
+
+    await Promise.all(
+      this._requests.map(r => Promise.resolve(r.promise).then(NOOP, NOOP))
```
Comment (jasonmit) [inline packages/@pollyjs/core/src/polly.js]: Whats the role of the noops here

File: packages/@pollyjs/utils/src/utils/deferred-promise.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Create a deferred promise with `resolve` and `reject` methods.
+ *
+ * @returns {Promise}
+ */
+export default class DeferredPromise extends Promise {
```
Comment (jasonmit) [inline packages/@pollyjs/utils/src/utils/deferred-promise.js]: Could be part of core since it's only used there

File: tests/integration/adapter-tests.js [hunk 4]
```diff
@@ -99,4 +99,30 @@ export default function adapterTests() {
...
+      res.sendStatus(200);
+    });
+
+    this.fetchRecord();
+    this.fetchRecord();
+    this.fetchRecord();
+
+    // Wait for all requests to be made
+    // (Puppeteer adapter takes some time to make the actual request)
+    while (requests.length !== 3) {
```
Comment (jasonmit) [inline tests/integration/adapter-tests.js]: Is this only noticed within the puppeteer adapter?  Worry we could mask potential future bugs by doing this always in all the adapter tests.

Comment (offirgolan) [reply tests/integration/adapter-tests.js]: > Is this only noticed within the puppeteer adapter?

Yes.

We use these tests in many places for both node and browser. I'm not really sure whats a better way to do this while making sure all adapters are tested for.

Comment (offirgolan) [reply tests/integration/adapter-tests.js]: > I wonder also if this could be mitigated by resolving the flush promise in the next frame.

What do you mean by this?

Comment (jasonmit) [reply tests/integration/adapter-tests.js]: > We use these tests in many places for both node and browser. I'm not really sure whats a better way to do this while making sure all adapters are tested for.

Skip the test in puppeteer.  Then create a test with this while loop within puppeteer's test suite so that we don't mask future bugs on the other adapters. 

> I wonder also if this could be mitigated by resolving the flush promise in the next frame.

Lets chat, I'd like to know more about what is going on here.  Specifically "takes some time".  Do we know how much time?  Does this while loop loop more than once?

File: packages/@pollyjs/adapter-puppeteer/tests/integration/adapter-test.js [hunk 5]
```diff
@@ -64,4 +64,40 @@ describe('Integration | Puppeteer Adapter', function() {
...
+      })
+      .on('request', req => requests.push(req));
+
+    this.page.on('requestfinished', () => resolved.push(i++));
+
+    this.fetchRecord();
+    this.fetchRecord();
+    this.fetchRecord();
+
+    // Since it takes time for Puppeteer to execute the request in the browser's
```
[the remark this answers is not in the document]
Comment (offirgolan) [reply packages/@pollyjs/adapter-puppeteer/tests/integration/adapter-test.js]: As discussed, there are no currently known edge cases. This is a test nuance that only relates to this test scenario.
